src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,128 @@
+package bio.terra.common.retry;
+
+import static java.time.Instant.now;
+
+import bio.terra.common.exception.InternalServerErrorException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. All constants are hard-coded, as
+ * Sam failures are generally just short periods of down-time.
+ *
+ * <p>SamRetry throws either the underlying ApiException from Sam or an InterruptedException.
+ */
+public class SamRetry {
+  private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+
+  // The retry function starts with INITIAL_WAIT between retries, and doubles that until it
+  // reaches MAXIMUM_WAIT, after which all retries are MAXIMUM_WAIT apart.
+  private static final Duration MAXIMUM_WAIT = Duration.ofSeconds(30);
+  private static final Duration INITIAL_WAIT = Duration.ofSeconds(10);
+  private static final Duration OPERATION_TIMEOUT = Duration.ofSeconds(300);
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  SamRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  protected SamRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface SamVoidFunction {
+    void apply() throws ApiException, InterruptedException;
```

Comment:
With all this said, I think the main thing is that `retry` throws an `InterruptedException`, whereas the SamFunction and SamVoidFunction interfaces don't need to be cause the underlying client library won't throw InterruptedExceptions. But leaving `throws InterruptedException` on the interface also doesn't really hurt anyone.

Reply:
Ah OK. Well I will leave it for now then.